chain-signatures/node/src/protocol/signature.rs

```diff
@@ -28,34 +27,91 @@ use tokio::sync::mpsc::error::TryRecvError;
 use tokio::sync::{mpsc, RwLock};
 
 use near_account_id::AccountId;
-use near_primitives::hash::CryptoHash;
-
-pub type ReceiptId = near_primitives::hash::CryptoHash;
 
 /// This is the maximum amount of sign requests that we can accept in the network.
 const MAX_SIGN_REQUESTS: usize = 1024;
 
-pub struct SignRequest {
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct SignId {
+    #[serde(with = "serde_bytes")]
     pub request_id: [u8; 32],
-    pub request: ContractSignRequest,
+    #[serde(with = "cbor_scalar")]
+    pub payload: Scalar,
+    #[serde(with = "cbor_scalar")]
     pub epsilon: Scalar,
+}
+
+impl std::hash::Hash for SignId {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.request_id.hash(state);
+        self.payload.to_bytes().hash(state);
+        self.epsilon.to_bytes().hash(state);
+    }
+}
+
+impl std::fmt::Debug for SignId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("SignId")
+            .field(&near_primitives::hash::CryptoHash(self.request_id))
+            .field(&hex::encode(self.payload.to_bytes()))
+            .field(&hex::encode(self.epsilon.to_bytes()))
+            .finish()
+    }
+}
+
+impl SignId {
+    pub fn new(request_id: [u8; 32], epsilon: Scalar, payload: Scalar) -> Self {
+        Self {
+            request_id,
+            epsilon,
+            payload,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct SignArgs {
+    #[serde(with = "serde_bytes")]
     pub entropy: [u8; 32],
-    pub time_added: Instant,
+    pub path: String,
+    pub key_version: u32,
+}
+
+/// All relevant info pertaining to an Indexed sign request from an indexer.
+#[derive(Debug, Clone, PartialEq)]
+pub struct IndexedSignRequest {
+    pub id: SignId,
+    pub args: SignArgs,
+    pub chain: Chain,
+    pub timestamp: Instant,
+}
+
+/// The sign request for the node to process. This contains relevant info for the node
+/// to generate a signature such as what has been indexed and what the node needs to maintain
+/// metadata-wise to generate the signature.
+#[derive(Debug, Clone, PartialEq)]
+pub struct SignRequest {
+    pub indexed: IndexedSignRequest,
+    pub proposer: Participant,
+    pub participants: Vec<Participant>,
 }
 
 pub struct SignQueue {
     me: Participant,
-    sign_rx: Arc<RwLock<mpsc::Receiver<SignRequest>>>,
-    my_requests: VecDeque<(SignRequest, Participant, Vec<Participant>)>,
-    other_requests: HashMap<SignRequestIdentifier, (SignRequest, Participant, Vec<Participant>)>,
+    sign_rx: Arc<RwLock<mpsc::Receiver<IndexedSignRequest>>>,
+    my_requests: VecDeque<SignRequest>,
+    other_requests: HashMap<SignId, SignRequest>,
```

Comment:
Here we have a map for SignId -> SignRequest, but SignRequest has IndexedSignRequestand it has the same SignId.

Reply:
it doesn't matter if we store them twice, because it's better if we can easily access the request instead of iterating through the whole list of requests